flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunction.java

```diff
@@ -16,99 +16,110 @@
  * limitations under the License.
  */
 
-package org.apache.flink.table.runtime.functions;
+package org.apache.flink.table.runtime.functions.table;
 
 import org.apache.flink.annotation.Internal;
-import org.apache.flink.table.catalog.DataTypeFactory;
+import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.data.ArrayData;
 import org.apache.flink.table.data.GenericRowData;
 import org.apache.flink.table.data.MapData;
-import org.apache.flink.table.functions.TableFunction;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.UserDefinedFunction;
+import org.apache.flink.table.runtime.functions.BuiltInSpecializedFunction;
 import org.apache.flink.table.types.DataType;
-import org.apache.flink.table.types.inference.TypeInference;
-import org.apache.flink.table.types.inference.TypeStrategies;
 import org.apache.flink.table.types.logical.ArrayType;
 import org.apache.flink.table.types.logical.LogicalType;
 import org.apache.flink.table.types.logical.MapType;
 import org.apache.flink.table.types.logical.MultisetType;
 import org.apache.flink.table.types.logical.RowType;
-import org.apache.flink.table.types.utils.DataTypeUtils;
 
-import static org.apache.flink.table.types.logical.LogicalTypeRoot.ROW;
-import static org.apache.flink.table.types.logical.LogicalTypeRoot.STRUCTURED_TYPE;
-
-/** Utility for handling SQL's {@code UNNEST}. */
+/**
+ * Flattens ARRAY, MAP, and MULTISET using a table function. It does this by another level of
+ * specialization using a subclass of {@link UnnestTableFunctionBase}.
+ */
 @Internal
-public final class SqlUnnestUtils {
+public class UnnestRowsFunction extends BuiltInSpecializedFunction {
+
+    public UnnestRowsFunction() {
+        super(BuiltInFunctionDefinitions.INTERNAL_UNNEST_ROWS);
+    }
 
-    public static UnnestTableFunction createUnnestFunction(LogicalType t) {
-        switch (t.getTypeRoot()) {
+    @Override
+    public UserDefinedFunction specialize(SpecializedContext context) {
+        final LogicalType argType =
+                context.getCallContext().getArgumentDataTypes().get(0).getLogicalType();
+        switch (argType.getTypeRoot()) {
             case ARRAY:
-                final ArrayType arrayType = (ArrayType) t;
+                final ArrayType arrayType = (ArrayType) argType;
                 return new CollectionUnnestTableFunction(
-                        arrayType,
-                        arrayType.getElementType(),
+                        context,
+                        getUnnestedType(arrayType),
```

Comment:
Can you replace this method call here with the actual logic? e.g. here becomes `arrayType.getElementType()`. It's easier to read and helps understanding this method